galsim/real.py

```diff
@@ -641,13 +673,23 @@ def getIndexForID(self, id):
         else:
             raise GalSimValueError('ID not found in list of IDs',id, self.ident)
 
+    def _maybe_preload(self):
+        # Preload all files if desired.
+        # This is delayed until the first time we might need it, since we might only need
+        # to know nobjects and not load the data at all.  The first time we try to do something
+        # that needs the files, we'll call preload (if requested).
+        if self._preload:
+            self.preload()
+            self._preload = False  # Once we've loaded them.  Don't do it again.
+
     def preload(self):
         """Preload the files into memory.
 
         There are memory implications to this, so we don't do this by default.  However, it can be
         a big speedup if memory isn't an issue.
         """
         from ._pyfits import pyfits
+        self.loaded_lock.acquire()
```

Comment:
I see a comment later in this file that the context manager syntax used to not work?  I definitely used in in the [phase_screens stuff](https://github.com/GalSim-developers/GalSim/blob/releases/2.2/galsim/phase_screens.py#L343) though, so maybe that means now it's safe to switch?

Reply:
Looks like.  I don't really remember the context where I had problems before.  That comment stems from a commit in 2013 (e1d51a4f53e985).  So could even have been a py2.6 problem.  I switched them all to use the `with lock:` syntax.  Everything seems to work fine on my machine for both 2.7 and 3.7.